### ISSUE TITLE
select.form-control and input[type=text].form-control are different heights

### ISSUE THREAD

**Issue by cvrebert** [twbs/bootstrap#17194]

**Comment by arthurbath** [comment]
Selects are a pain. Safari renders them with an undesirable background gradient, and Firefox attaches that awful Win98-esque dropdown button. But if you set `appearance: none`, they lose their up-down arrow indicators.

Here's an extension for `select.form-control` that I've been using that strips the appearance, then adds the dropdown indicator back via a base64 SVG (courtesy [Entypo](http://www.entypo.com)):

    select.form-control {
      -moz-appearance: none;
      -webkit-appearance: none;
      background-image: url(data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiIHN0YW5kYWxvbmU9Im5vIj8+PHN2ZyB3aWR0aD0iMTBweCIgaGVpZ2h0PSIxOHB4IiB2aWV3Qm94PSIwIDAgMTAgMTgiIHZlcnNpb249IjEuMSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB4bWxuczp4bGluaz0iaHR0cDovL3d3dy53My5vcmcvMTk5OS94bGluayIgeG1sbnM6c2tldGNoPSJodHRwOi8vd3d3LmJvaGVtaWFuY29kaW5nLmNvbS9za2V0Y2gvbnMiPiAgICAgICAgPHRpdGxlPlVudGl0bGVkPC90aXRsZT4gICAgPGRlc2M+Q3JlYXRlZCB3aXRoIFNrZXRjaC48L2Rlc2M+ICAgIDxkZWZzPjwvZGVmcz4gICAgPGcgaWQ9IlBhZ2UtMSIgc3Ryb2tlPSJub25lIiBzdHJva2Utd2lkdGg9IjEiIGZpbGw9Im5vbmUiIGZpbGwtcnVsZT0iZXZlbm9kZCIgc2tldGNoOnR5cGU9Ik1TUGFnZSI+ICAgICAgICA8ZyBpZD0ic2VsZWN0LWFycm93cyIgc2tldGNoOnR5cGU9Ik1TTGF5ZXJHcm91cCIgZmlsbD0iIzAwMDAwMCI+ICAgICAgICAgICAgPHBhdGggZD0iTTUsMCBMMCw3IEwxMCw3IEw1LDAgTDUsMCBaIE01LDE4IEwxMCwxMSBMMCwxMSBMNSwxOCBMNSwxOCBaIiBpZD0iU2hhcGUiIHNrZXRjaDp0eXBlPSJNU1NoYXBlR3JvdXAiPjwvcGF0aD4gICAgICAgIDwvZz4gICAgPC9nPjwvc3ZnPg==);
      background-size: .5em;
      background-repeat: no-repeat;
      background-position: right .5em center;
      padding-right: 2em;
      overflow: hidden;
      text-overflow: ellipsis;
    }

Got a little frustrated by the text abruptly cutting off at the padding line, so I also added the `text-overflow: ellipsis`, too, though it's not entirely necessary. Not sure if there are problems with this solution (screen readers?) that would make it unsuitable for official inclusion.

*Chrome*
<img width="812" alt="chrome--after" src="https://cloud.githubusercontent.com/assets/3257528/9453179/75577218-4a6d-11e5-9e93-21aac11039da.png">
*Chrome (input-sm)*
<img width="813" alt="chrome--after--sm" src="https://cloud.githubusercontent.com/assets/3257528/9453187/758ae332-4a6d-11e5-920c-12aad24f962a.png">
*Firefox*
<img width="810" alt="firefox--after" src="https://cloud.githubusercontent.com/assets/3257528/9453185/758909d6-4a6d-11e5-940a-6019334e43c0.png">
*Safari*
<img width="668" alt="safari--after" src="https://cloud.githubusercontent.com/assets/3257528/9453184/75896a02-4a6d-11e5-8b30-41144854a209.png">

**Comment by cvrebert** [comment]
@arthurbath What you're proposing is something more appropriate for the completely custom form controls (http://v4-alpha.getbootstrap.com/components/forms/#select-menu ), not our vanilla `.form-control`, which is supposed to be native-ish (well, as native as the browser engine itself allows).

**Comment by cvrebert** [comment]
CC: @mdo

**Comment by mdo** [comment]
Fixed in caa36b8.

**Comment by cvrebert** [comment]
@mdo Still not quite right (OS X Chrome):
<img width="895" alt="select-height" src="https://cloud.githubusercontent.com/assets/419884/12879196/8e2db130-cde4-11e5-9aa5-fcba0f62bfcf.png">

**Comment by eliottrobson** [comment]
Is this a similar issue to #18786 ?

**Comment by cvrebert** [comment]
@eliottrobson Only slightly.

**Comment by eliottrobson** [comment]
@cvrebert when I was looking at that issue I did notice that my changes did alter the height of the select element. But the way I did it isn't ideal, hard-coded heights etc... Is there any way we can apply the same logic but calculate the values?

**Comment by torbjoernk** [comment]
I might miss something, but does the addition introduced by [`caa36b8`](https://github.com/twbs/bootstrap/commit/caa36b8dd723945a788b1dcb2c845ee05532155e) have to be extended/duplicated for `.form-control-sm` and `.form-control-lg` with `height: $input-height-sm;` and `height: $input-height-lg;` respectively?

**Comment by cvrebert** [comment]
@torbjoernk Quite possibly. See #19771.

**Comment by leetskills** [comment]
I found that the issue for me was the top & bottom border widths being added to the calculated height of text form controls, but not being part of the `$input-height` calculation. Would something like the below be a valid fix?

```sass
$input-height: (($font-size-base * $input-line-height) + ($input-padding-y * 2)) + ($input-btn-border-width * 2);
```

Note: in order to do this, I had to set `$border-width` to `.0625rem` (equivalent to 1px based on 16px `$font-size-base`.

This can be extended to the `-sm/-lg` variations but the heights are still off as the calculated line height doesn't round to a whole number.
